robot/Cumulus/tests/browser/learning_robot.robot

```diff
@@ -12,48 +11,81 @@ Suite Teardown  Delete Records and Close Browser
         Populate Modal Form
         ...                                  Opportunity Name=${opportunity_name}
-        ...                                  Account Name=${data}[contact][LastName] Household
+        ...                                  Account Name=${account_name}
         ...                                  Amount=${amount}
-        ...                                  Primary Campaign Source=${campaign}[Name]
+        ...                                  Primary Campaign Source=${campaign_name}
         Select Value From Dropdown           Stage      ${stage_name}
         Open Date Picker                     Close Date
         Pick Date                            Today
         Click Modal Button                   Save 
+        Wait Until Modal Is Closed
+        ${opp_id}=                           Save Current Record Id For Deletion                Opportunity
+        [return]                             ${opp_id}
+
+API Create Campaign GAU Allocation
+        [Arguments]     ${gau_id}       ${campaign_id}          &{fields}
+        ${ns}=                               Get Npsp Namespace Prefix
+        ${all_id}=                           Salesforce Insert                   
+        ...                                  ${ns}Allocation__c
+        ...                                  ${ns}General_Accounting_Unit__c=${gau_id}
+        ...                                  ${ns}Campaign__c=${campaign_id}
+        ...                                  &{fields}
+        &{gau_alloc}=                        Salesforce Get  ${ns}Allocation__c  ${all_id}
+        [return]                             &{gau_alloc} 
+
+Browser Create Campaign GAU Allocation
+        [Arguments]     ${campaign_name}
+        Select Tab                           Related
+        Click Object Button                  New
+        Populate Modal Form                  Percent=100
+        ...                                  Campaign=${campaign_name}
+        Click Modal Button                   Save
+
+Verify GAU Allocation is Automatically Created
         Select Tab                           Related    
-        Validate Related Record Count        GAU Allocations         1                                
\ No newline at end of file
+        Validate Related Record Count        GAU Allocations         1        
+
+
+*** Variables ***    
+&{contact1_fields}       Email=test@example.com
+${amount}                100
+${stage_name}            Closed Won
+
+
+*** Test Cases ***
+Assign GAU to Campaign and Verify Allocation on Opportunity
+        ${gau_id}=                           Create GAU                          
+        API Create Campaign GAU Allocation   ${gau_id}
+        ...                                  &{campaign}[Id]
+        ...                                  ${ns}Percent__c=100.0 
+        Create Opportunity With Campaign     ${data}[contact][LastName] Household  
+        ...                                  ${campaign}[Name]
```

Comment:
Any reason this needs to be UI? If you just testing the trigger logic, you might as well just create Opportunity through API. Looks like you can use this existing Keyword: https://github.com/SalesforceFoundation/NPSP/blob/master/robot/Cumulus/resources/NPSP.robot#L50

Reply:
@Julian88Tex I felt like this should try to re-create the user scenario as close as possible through the UI, but would you recommend API?

Reply:
@sungysang it all depends on what you're trying to test. If you trying to test the trigger, then API should be fine IMO since I wouldn't expect UI to have any impact on trigger logic vs. making an API call. If you're trying to test UI then stick with UI.